src/mame/brother/lw840.cpp

```diff
@@ -0,0 +1,362 @@
+		io_kbrow(*this, "kbrow.%u", 0),
+		rom(*this, "maincpu"),
+		sram(*this, "sram")
+	{ }
+
+	void lw840(machine_config& config);
+
+protected:
+	// driver_device overrides
+	void machine_start() override;
+	void machine_reset() override;
+	void video_start() override;
+
+private:
+	// devices
+	required_device<h83003_device> maincpu;
+	required_device<screen_device> screen;
+	required_device<gm82c765b_device> fdc;
+	required_device<beep_device> beeper;
+	required_ioport_array<9> io_kbrow;
+	required_region_ptr<uint16_t> rom;
+	required_shared_ptr<uint16_t> sram;
+
+	uint8_t keyboard{};
+	uint8_t irq_toggle{};
+
+	// screen updates
+	uint32_t screen_update(screen_device& screen, bitmap_rgb32& bitmap, const rectangle& cliprect);
+
+	uint16_t port7_r(offs_t, uint16_t mem_mask = ~0) {
+		auto row = keyboard & 0x0f;
+		if(row <= 8)
+			return io_kbrow[row]->read();
+
+		// seems to be able to control power-on self test if not 0xff
+		return 0xffff & mem_mask;
+	}
+	uint16_t keyboard_r(offs_t, uint16_t mem_mask = ~0) {
+		return (keyboard << 8) & mem_mask; 
+	}
+	void keyboard_w(uint16_t data) { 
+		keyboard = data >> 8; 
+	}
+	TIMER_DEVICE_CALLBACK_MEMBER(int2_timer_callback) {
+		irq_toggle = ~irq_toggle;
+		maincpu->set_input_line(INPUT_LINE_IRQ2, irq_toggle ? ASSERT_LINE : CLEAR_LINE);
+	}
+
+	static void floppy_formats(format_registration& fr) {
+		fr.add(FLOPPY_PC_FORMAT);
+	}
+
+	DECLARE_WRITE_LINE_MEMBER(fdc_interrupt) {
+		maincpu->set_input_line(INPUT_LINE_IRQ4, state ? ASSERT_LINE : CLEAR_LINE);
+	}
+	DECLARE_WRITE_LINE_MEMBER(fdc_drq) {
+		maincpu->set_input_line(H8_INPUT_LINE_DREQ0, state ? ASSERT_LINE : CLEAR_LINE);
+	}
+	uint16_t disk_inserted_r(offs_t, uint16_t mem_mask = ~0) {
+		// bit#6: disk inserted
+		// bit#7: ??1.44mb
+		// bit#1: ???
+		// bit#0: ???
+
+		// hack: always disk inserted, HD-disk
+		return (0b01000000 << 8) & mem_mask;
+	}
+
+	void map_program(address_map& map) {
+		map(0x000000, 0x3fffff).rom();
+		map(0x5f8000, 0x5fffff).ram().share("sram"); // SRAM
+		map(0x600000, 0x67ffff).ram(); // DRAM
+		map(0xe00000, 0xe00007).m(fdc, FUNC(gm82c765b_device::map));
+		map(0xe00030, 0xe00041).noprw(); // just to shut up the error.log
+		map(0xec0000, 0xec0001).r(FUNC(lw840_state::keyboard_r)).w(FUNC(lw840_state::keyboard_w));
+		map(0xec0004, 0xec0005).r(FUNC(lw840_state::disk_inserted_r));
+	}
+
+	void map_io(address_map& map) {
+		map(h8_device::PORT_7, h8_device::PORT_7).r(FUNC(lw840_state::port7_r));
+	}
+};
+
+void lw840_state::video_start()
+{
+}
+
+uint32_t lw840_state::screen_update(screen_device& screen, bitmap_rgb32& bitmap, const rectangle& cliprect)
+{
+	const rgb_t palette[]{
+		0xffffffff,
+		0xff000000,
+	};
+
+	auto vram = reinterpret_cast<uint16_t*>(&sram[0x300 / sizeof(uint16_t)]);
+
+	for(auto y = 0; y < 400; y++) {
+		uint32_t* p = &bitmap.pix(y);
+		for(auto x = 0; x < 640; x += 16) {
+			auto gfx = vram[(y * 640 + x) / 16];
+			for(int i = 15; i >= 0; i--)
+				*p++ = palette[BIT(gfx, i)];
+		}
+	}
```

Comment:
This won’t work properly on big Endian hosts.  You need to either adapt it to work with bytes, or use `little_endianize_int16` on the data.  You should also at least restrict the rows you update to the range `cliprect.top()` to `cliprect.bottom()`, inclusive.

Reply:
Actually, it's already a `uint16_t` pointer – why do you have the `reinterpret_cast`?

Reply:
my bad ;) forgot after I changed something else